### PR TITLE
RRGraphView node_ptc_num(), node_pin_num(), node_class_num(), node_track_num() Implementation

### PR DESCRIPTION
**Pull request by ethanroj23**
### Description

In this PR, I have implemented `RRGraphView::node_ptc_num()`, `RRGraphView::node_pin_num()`, `RRGraphView::node_class_num()`, `RRGraphView::node_track_num()` throughout VTR. Every time `device_ctx.rr_nodes[node].node_ptc_num()` was called has been replaced with `rr_graph.node_ptc_num(RRNodeId(node))`. The same pattern was followed for the other functions in this PR. In order to do this, I followed a pattern similar to that in the last PR.

### Motivation and Context

These changes are a continuation of the RRGraphView API implementation effort.

### How Has This Been Tested?

I have run the regression tests for `vtr_reg_strong` along with the QoR testing found at `    $ ../scripts/run_vtr_task.py regression_tests/vtr_reg_nightly_test3/vtr_reg_qor_chain`. Results from these QoR tests can be found below. The file containing all results can be found [here](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/6555913/rr_graph_view_node_type.xlsx).

Only rows that are different are shown. For some reason, `vtr_flow_elapsed_time` was empty in the spreadsheet.

|	| VTR Before These Changes	|With Changes in this PR|
|---|---|---|
|odin_synth_time	|1|	0.984647269|
|abc_synth_time	|1|	1.001007005|
|max_vpr_mem	|1|	0.999800843|
|pack_time	|1|	0.997122006|
|place_time	|1|	0.992015738|
|min_chan_width_route_time	|1|	0.996755548|
|crit_path_route_time	|1|	1.01858794|

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: vpr/src/device/rr_graph_view.h
Comment (tangxifan) [inline vpr/src/device/rr_graph_view.h:252]: The ``node_ptc_num()`` requires additional comments to guide developers. Can you adapt the text in RRGraphBuilder in this context? 

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/bae983f4e018a96b92a8a0679c27e75d528e78b3/vpr/src/device/rr_graph_builder.h#L72-L99

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2196]: Since this is in the CHANX branch, we should use the node_track_num() API. Try to avoid using ``node_ptc_num()`` unless we have to

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2199]: Same issue. Use the ``node_track_num()`` API.

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2204]: Same issue. Use the ``node_track_num()`` API.

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2308]: Same issue. Use the ``node_pin_num()`` API.

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2367]: Same issue. Use the ``node_class_num()`` API.

File: vpr/src/draw/draw.cpp
Comment (tangxifan) [inline vpr/src/draw/draw.cpp:2623]: Same issue. Use the ``node_track_num()`` API.

File: vpr/src/route/check_route.cpp
Comment (tangxifan) [inline vpr/src/route/check_route.cpp:217]: Same issue. Use the ``node_class_num()`` API.

File: vpr/src/draw/search_bar.cpp
Comment (tangxifan) [inline vpr/src/draw/search_bar.cpp:195]: Same issue. Use the ``node_pin_num()`` API.

**Review by tangxifan — CHANGES_REQUESTED** [review]
@ethanroj23 Just fix needed. 
- Add proper comments for the APIs
- Recheck the use of the APIs in client functions. When the node_type is specified, we should not use ``node_ptc_num()``. I have suggested some code changes. But I have not covered all the code changes. Can you double check all the locations in client functions? If you find ``node_ptc_num()`` is not used. My opinion is that we can simply delete this API. @vaughnbetz Let me know your opinion on the ``node_ptc_num()`` API

File: vpr/src/draw/search_bar.cpp
Comment (tangxifan) [reply vpr/src/draw/search_bar.cpp:195]: @ethanroj23 May be it is a typo. Should we use ``node_pin_num()`` here?

File: vpr/src/route/check_route.cpp
Comment (tangxifan) [inline vpr/src/route/check_route.cpp:341]: I read the codes in this function. It seems that the ``to_ptc`` is only used by SOURCE and IPIN nodes. Should we use specific API in these cases rather than depositing here? I do not have a strong opinion here. Your call.

File: vpr/src/route/clock_connection_builders.cpp
Comment (tangxifan) [inline vpr/src/route/clock_connection_builders.cpp:124]: Here the node_type is explicitly defined as SINK (See LINE 110). We can use the ``node_class_num()`` API here.

**Review by tangxifan — CHANGES_REQUESTED** [review]
@ethanroj23 Thanks for the effort. A few more changes on the APIs required. I do not have further comments then. We can discuss this PR in the upcoming weekly meeting.

File: vpr/src/route/check_route.cpp
Comment (ethanroj23) [reply vpr/src/route/check_route.cpp:341]: I think the code is fine this way. We could move the call lower into the individual case statements, thought if you think that would be better.

Comment (tangxifan) [reply vpr/src/route/check_route.cpp:341]: No worries. Let's keep it in this way.

**Comment by ethanroj23** [comment]
@tangxifan Do you know why this PR is still pending? It looks like the "Gitpod" test has been hung for a while. Perhaps it needs to be rebuilt entirely?

**Comment by tangxifan** [comment]
@ethanroj23 Gitpod glitches from time to time. I see such time happened once. I do not think there is an issue.

**Comment by ethanroj23** [comment]
> @ethanroj23 Gitpod glitches from time to time. I see such time happened once. I do not think there is an issue.

Ok sounds good. Is there anything else that needs to be done for this PR before we merge it?

**Review by tangxifan — APPROVED** [review]

**Comment by tangxifan** [comment]
> > @ethanroj23 Gitpod glitches from time to time. I see such time happened once. I do not think there is an issue.
> 
> Ok sounds good. Is there anything else that needs to be done for this PR before we merge it?

Look good to me. See if @vaughnbetz has any comments

**Comment by tangxifan** [comment]
@ethanroj23 @vaughnbetz As discussed last week, merge this PR. We can quickly follow-up if any issues are seen.

**Pull request merged**
